src/smt/optimization_solver.h

```diff
@@ -80,38 +85,163 @@ class Objective
+  virtual ~OMTOptimizerInteger() = default;
+  /** Minimize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> minimize(SmtEngine* parentSMTSolver,
+                                              Node target) override;
+  /** Maximize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> maximize(SmtEngine* parentSMTSolver,
+                                              Node target) override;
+
+ private:
+  /** Handles the optimization query specified by objType
+   * (=OBJECTIVE_MINIMIZE/MAXIMIZE) **/
+  std::pair<OptResult, Node> optimize(SmtEngine* parentSMTSolver,
+                                      Node target,
+                                      ObjectiveType objType);
+};
+
+/**
+ * Optimizer for BitVector type
+ */
+class OMTOptimizerBitVector : public OMTOptimizer
+{
+ public:
+  OMTOptimizerBitVector(bool isSigned);
+  virtual ~OMTOptimizerBitVector() = default;
+  /** Minimize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> minimize(SmtEngine* parentSMTSolver,
+                                              Node target) override;
+  /** Maximize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> maximize(SmtEngine* parentSMTSolver,
+                                              Node target) override;
+
+ private:
+  /** Computes the BitVector version of (a + b) / 2 without overflow,
+   * rounding towards -infinity: -1.5 --> -2 and 1.5 --> 1
+   * same as the rounding scheme for int32_t
+   **/
+  BitVector computeAverage(const BitVector& a,
+                           const BitVector& b,
+                           bool isSigned);
+  /** Initialize an SMT subsolver for offline optimization purpose **/
+  std::unique_ptr<SmtEngine> initOptChecker(SmtEngine* parentSMTSolver);
```

Comment:
plz document why this is needed for the bv sub-class but not for the integer one.

Reply:
it could have it...

Reply:
added